### PR TITLE
feat!: collect perf results from outside oclif

### PR DESCRIPTION
**Pull request by mshanemc**
Allows plugins/libraries to consume Performance.

BREAKING CHANGE: Performance.mark(),  Perforance.getResult() now requires an `owner` property meant to be the npm package name.

has a special getter for `oclifPerf` 
(OclifHighlights, which does most of the transformations formerly in plugin-telemetry)

---

QA:  this is painful, sorry.

check out this branch and `yarn link`
in your local CLI directly, `yarn link @oclif/core`

You should be able to run commands with `DEBUG=oclif-perf` and see the oclif output.  There's nothing going to AppInsights OR any non-oclif results.

if you want to validate what goes into AppInsights,
yarn link this branch into plugin-telemetry using this PR https://github.com/salesforcecli/plugin-telemetry/pull/515
sf plugins link [plugin-telemetry]
then run some command with `SF_TELEMETRY_DEBUG=true DEBUG=sfdx:telemetry*` so you can get your CLIId to use for searching appInsights logs

Oh, you actually want to try the feature (peformance marks outside of oclif) ??
It's easiest to put these in a linked plugin (ex: PDR).  You have to use plugin linked into sf so that you get all the hooks (`bin/dev` won't do)
I have some already created for SDR in https://github.com/forcedotcom/source-tracking/pull/478, but you'd need to yarn link oclif/core into it and then yarn link that into your linked plugin

@W-14182563@

### CODE REVIEW

File: src/performance.ts [hunk 1]
```diff
@@ -12,17 +12,21 @@ type PerfResult = {
 }
 
 type PerfHighlights = {
-  configLoadTime: number;
```
Comment (mshanemc) [inline src/performance.ts:15]: these will now be pre-formatted for AppInsights, which doesn't like nested obj.

File: src/performance.ts [hunk 2]
```diff
@@ -12,17 +12,21 @@ type PerfResult = {
...
+  'oclif.initMs': number;
+  'oclif.commandLoadMs': number;
+  'oclif.initHookMs': number;
+  'oclif.prerunHookMs': number;
+  'oclif.postrunHookMs': number;
+  'oclif.commandRunMs': number;
+  'oclif.corePluginsLoadMs': number;
+  'oclif.userPluginsLoadMs': number;
+  'oclif.linkedPluginsLoadMs': number;
   pluginLoadTimes: Record<string, {duration: number, details: Details}>;
```
Comment (mshanemc) [inline src/performance.ts:26]: telemetry.record will skip object types and not record them to the telemetry file

File: src/performance.ts [hunk 3]
```diff
@@ -202,33 +215,46 @@ export class Performance {
   public static debug(): void {
     if (!Performance.enabled) return
 
-    const debug = require('debug')('perf')
-    debug('Total Time: %sms', Performance.highlights.runTime.toFixed(4))
```
Comment (mshanemc) [inline src/performance.ts:206]: this was creating highlights and then letting plugin-telemetry re-massage the data.  Now we only do it once.

File: src/performance.ts [hunk 4]
```diff
@@ -202,33 +215,46 @@ export class Performance {
...
+        oclifDebug(`    ${plugin}: ${duration.toFixed(4)}ms`)
       }
     }
 
-    debug('Command Load Time: %sms', Performance.highlights.commandLoadTime.toFixed(4))
-    debug('Command Run Time: %sms', Performance.highlights.commandRunTime.toFixed(4))
+    oclifDebug('Command Load Time: %sms', Performance.oclifPerf['oclif.commandLoadMs'].toFixed(4))
+    oclifDebug('Command Run Time: %sms', Performance.oclifPerf['oclif.commandRunMs'].toFixed(4))
+
+    const nonOclifDebug = require('debug')('non-oclif-perf')
```
Comment (mshanemc) [inline src/performance.ts:247]: separate debug makes it easier to only get the perf stuff you want.

File: src/config/config.ts [hunk 5]
```diff
@@ -9,9 +9,9 @@ import {URL, fileURLToPath} from 'node:url'
 import {arch, userInfo as osUserInfo, release, tmpdir, type} from 'node:os'
 import {compact, ensureArgObject, getHomeDir, getPlatform, isProd, requireJson} from '../util'
 import {join, sep} from 'node:path'
-
+// eslint-disable-next-line sort-imports
```
Comment (mdonnalley) [inline src/config/config.ts:12]: Why disable?

File: src/performance.ts (hunk 4 above)
Comment (mdonnalley) [inline src/performance.ts:247]: This is a misleading name. What if we end up putting a bunch of markers in `@oclif/plugin-update`?

File: src/performance.ts [hunk 6]
```diff
@@ -202,33 +215,46 @@ export class Performance {
...
+
+    const nonOclifPerf = Performance.results
+    if (nonOclifPerf.size > 0) {
+      nonOclifDebug('Non-Oclif Performance Measurements:')
```
Comment (mdonnalley) [inline src/performance.ts:251]: I think it'd be more accurate to say `Non-Core Performance Measurements:`

File: src/config/config.ts (hunk 5 above)
Comment (mshanemc) [reply src/config/config.ts:12]: Try it without.  There's nothing that make it happy

File: src/performance.ts (hunk 6 above)
Comment (mshanemc) [reply src/performance.ts:251]: Yeah, I'll rename that other one, too.

File: src/config/config.ts (hunk 5 above)
Comment (mshanemc) [reply src/config/config.ts:12]: ok, now it works happily.  Cool, nevermind.

**Review by mdonnalley — APPROVED** [review]

**Pull request merged**